onnxruntime/core/optimizer/free_dim_override_transformer.cc

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/logging/logging.h"
+#include "core/session/inference_session.h"
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/free_dim_override_transformer.h"
+
+using namespace ONNX_NAMESPACE;
+using namespace ::onnxruntime::common;
+namespace onnxruntime {
+
+static std::string ToLower(std::string s) {
+  std::transform(s.begin(), s.end(), s.begin(), [](char c) {
+    return static_cast<char>(::tolower(c));
+  });
+
+  return s;
+}
+
+/*explicit*/ FreeDimensionOverrideTransformer::FreeDimensionOverrideTransformer(gsl::span<const FreeDimensionOverride> overrides_to_apply)
+  : GraphTransformer("FreeDimensionOverrideTransformer") {
+  for (const auto& o : overrides_to_apply) {
+    // Convert to lowercase to perform case-insensitive comparisons later
+    std::string denotation = ToLower(o.dimension_denotation);
+
+    dimension_override_by_denotation_.emplace(denotation, o.dimension_override);
+  }
+}
+
+Status FreeDimensionOverrideTransformer::ApplyImpl(Graph& graph, bool& modified, int /*graph_level*/) const {
+  for (const onnxruntime::NodeArg* graph_input : graph.GetInputs()) {
+    // Get the current input's type and shape
```

Comment:
This won't include optional inputs (inputs with a backing initializers). If you need to process those as well use GetInputsIncludingInitializers.

Reply:
Constant initializers would always have static shapes (I would hope) so presumably we actually wouldn't want the free dimension override to apply to them.